### ISSUE TITLE
Action Scheduler conflicts with Shibboleth SP (IP address mismatch)

### ISSUE THREAD

**Issue by wursterje** [Prospress/action-scheduler#981]
We are using a Shibboleth Service Provider (SP) setup for authentication and experiencing random logouts for users. We have finally identified the cause of this issue.

The Shibboleth SP sets a cookie named "_shibsession_xyz," which is bound to the client's IP address. We have observed that WP_Async_Request.php utilizes the entire request cookies for internal calls.

https://github.com/woocommerce/action-scheduler/blob/5b17c9b4e2eb38494759b65b24e7918b6c3c3de7/lib/WP_Async_Request.php#L153-L155

https://github.com/woocommerce/action-scheduler/blob/5b17c9b4e2eb38494759b65b24e7918b6c3c3de7/lib/WP_Async_Request.php#L91-L94

This poses a problem because the internal calls have a different IP address compared to the client's IP, and the SP removes the session. As you can see in the SP logs:

```
2023-09-05 16:45:36 DEBUG Shibboleth.Apache [943] shib_check_user: mapped https://*****/wp-admin/admin-ajax.php?action=as_async_request_queue_runner&nonce=******* to default
2023-09-05 16:45:36 DEBUG Shibboleth.SessionCache [943] shib_check_user [default]: searching local cache for session (_2a058db4fc91c36e********)
2023-09-05 16:45:36 DEBUG Shibboleth.SessionCache [943] shib_check_user [default]: session found locally, validating it for use
2023-09-05 16:45:36 WARN Shibboleth.SessionCache [943] shib_check_user [default]: client address mismatch, client (100.96.5.140), session (***.***.***.***)
2023-09-05 16:45:36 DEBUG Shibboleth.SessionCache [943] shib_check_user [default]: deleting local copy of session (_2a058db4fc91c36e********)
2023-09-05 16:45:36 DEBUG Shibboleth.Listener [943] shib_check_user [default]: sending message (remove::StorageService::SessionCache)
2023-09-05 16:45:36 DEBUG Shibboleth.Listener [943] shib_check_user [default]: send completed, reading response message
2023-09-05 16:45:36 WARN Shibboleth.ServiceProvider [943] shib_check_user [default]: error during session lookup: Your IP address (100.96.5.140) does not match the address recorded at the time the session was established.
```

Even if you do not use WooCommerce, some plugins like WP Rocket or WP Mail SMTP make use of the Action Scheduler as a internal dependency.

To avoid this problem one possible solution could be to add a regex filter function and an option for removing specific cookies for the Action Scheduler calls.

**Comment by barryhughes** [comment]
Are you able to set up some custom code, perhaps living in a mu-plugin file, on the following model? This is just a starting point, you'd need to tweak accordingly.

```php
add_filter( 'as_async_request_queue_runner_post_args', function ( $args ) { 
	$args['cookies'] = []; // Modify as needed.
	return $args;
} );
```

**Comment by wursterje** [comment]
Yes, we can set up custom code. Using the following code, we can successfully remove the Shibboleth session cookie:

```
add_filter( 'as_async_request_queue_runner_post_args', function ( $args ) { 
	$args['cookies'] = array_filter($args['cookies'],
		function ($key) {
			return strpos($key, '_shibsession_') !== 0;
		},
		ARRAY_FILTER_USE_KEY
	);
	
	return $args;
} );
```

For reference: Another option would be to set consistentAddress to "false" in the shibboleth2.xml (https://shibboleth.atlassian.net/wiki/spaces/SP3/pages/2065334342/Sessions#Attributes). However, the downside is that it offers reduced security against cookie theft.

**Comment by barryhughes** [comment]
Thanks for sharing that.

I don't think there is any more we should be doing from within Action Scheduler in this case, so will close the issue (if I'm mistaken, though, please feel free to follow-up with further ideas).
